apps/src/storage/dataBrowser/dataVisualizer/Snapshot.jsx

```diff
@@ -4,28 +4,76 @@ import PropTypes from 'prop-types';
 import {connect} from 'react-redux';
 import moment from 'moment';
 import msg from '@cdo/locale';
+import {svgToDataURI} from '@cdo/apps/imageUtils';
 import {html2canvas} from '@cdo/apps/util/htmlToCanvasWrapper';
 import BaseDialog from '@cdo/apps/templates/BaseDialog';
 import PendingButton from '@cdo/apps/templates/PendingButton';
+import {ChartType} from '../dataUtils';
 import * as dataStyles from '../dataStyles';
 
+const INITIAL_STATE = {
+  isSnapshotOpen: false,
+  isCopyPending: false,
+  isSavePending: false,
+  imageSrc: require('./placeholder.png')
+};
+
 class Snapshot extends React.Component {
   static propTypes = {
+    chartType: PropTypes.number.isRequired,
     chartTitle: PropTypes.string.isRequired,
     selectedOptions: PropTypes.string.isRequired,
     // Provided via Redux
     tableName: PropTypes.string.isRequired,
     projectName: PropTypes.string.isRequired
   };
 
-  state = {
-    isSnapshotOpen: false,
-    isCopyPending: false,
-    isSavePending: false
+  state = {...INITIAL_STATE};
+
+  componentDidMount() {
+    this.isMounted_ = true;
+  }
+
+  componentWillUnmount() {
+    this.isMounted_ = false;
+  }
+
+  handleOpen = () => {
+    this.setState({isSnapshotOpen: true});
+    this.getImageFromChart();
   };
 
-  handleOpen = () => this.setState({isSnapshotOpen: true});
-  handleClose = () => this.setState({isSnapshotOpen: false});
+  handleClose = () => this.setState(INITIAL_STATE);
+
+  getImageFromChart = () => {
+    switch (this.props.chartType) {
+      case ChartType.BAR_CHART:
+      case ChartType.HISTOGRAM:
+      case ChartType.SCATTER_PLOT:
+        this.getImageFromGoogleChart();
+        break;
+      case ChartType.CROSS_TAB:
+        this.getImageFromCrossTab();
+        break;
+      default:
+    }
+  };
+
+  getImageFromCrossTab = () => {};
+
+  getImageFromGoogleChart = () => {
+    const container = document.getElementById('googleChartContainer');
+    const svgList = container && container.querySelectorAll('svg');
+    const svg = svgList && svgList[0];
+    if (!svg) {
+      return;
+    }
+    svgToDataURI(svg, 'image/png', {renderer: 'native'}).then(imageURI => {
+      if (this.isMounted_) {
```

Comment:
i don't think i understand why `this.isMounted_` is necessary -- won't this method only ever be called when the component has already mounted?

Reply:
i just noticed [Brad's comment](https://github.com/code-dot-org/code-dot-org/pull/33703#discussion_r395384136) that's basically the same -- feel free to answer there

Reply:
`getImageFromGoogleChart` will only be called when the component is mounted, but it's possible that the snapshot modal is closed before `svgToDataURI` finishes, in which case the component will be unmounted by the time the callback runs. That's why I have the `isMounted_` check before setting the state from the callback